### PR TITLE
Issue 911: [BigQuery] Using functions from other projects, quoted function names

### LINKED ISSUE #911: [BigQuery] Using functions from other projects
Hi there!

I've tested these 2 different ways of calling _remote_ functions and both worked inside the BigQuery editor.

```sql
SELECT
    `another-gcp-project`.functions.timestamp_parsing(log_tbl.first_move) AS first_move,
    `another-gcp-project.functions.timestamp_parsing`(log_tbl.orderdate) AS orderdate
FROM
    `gcp-project.data.year_2021` AS log_tbl
;
```

## Expected Behaviour

No errors.

## Observed Behaviour

```
L:   3 | P:  27 | L026 | Reference
                       | '`another-gcp-project`.functions.timestamp_parsing'
                       | refers to table/view 'functions' not found in the FROM
                       | clause or found in parent subquery.
L:   3 | P:  54 |  PRS | Found unparsable section: '(log_tbl.first_move) AS
                       | first_move'
L:   4 | P:   5 | L026 | Reference
                       | '`another-gcp-project.functions.timestamp_parsing`'
                       | refers to table/view 'functions' not found in the FROM
                       | clause or found in parent subquery.
L:   4 | P:  54 |  PRS | Found unparsable section: '(log_tbl.orderdate) AS
                       | orderdate'
L:   6 | P:  37 | L031 | Avoid using aliases in join condition
```

## Version

```
sqlfluff, version 0.4.1
```

## Configuration
```
[sqlfluff]
verbose = 0
dialect = bigquery
templater = python
exclude_rules = L009,L013

[sqlfluff:rules]
max_line_length = 120

[sqlfluff:templater:python:context]
source_table_id=my_table
start_date=456
end_date=456
fixed_carrier_id=456
fixed_destination_country_id=456
fixed_origin_country_id=456
number_of_buckets=123
bucket_id=123
fixed_carrier_ids=('1','2','3')
fixed_destination_country_ids=('1','2','3')
fixed_origin_country_ids=('1','2','3')
```

### PR DESCRIPTION
**Pull request by barrywhart**
Resolves #911

This PR changes the parse tree for `function_name`, hence the large number of files changed (`.yml` files for the parse test files).

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_ansi.py
Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_ansi.py:762]: Rather than allowing 0 or 1 names in front of the function name, allow any number. (BigQuery only uses 2, so we could limit it to 2 if that's preferable.)

File: src/sqlfluff/dialects/dialect_bigquery.py
Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_bigquery.py:82]: I had to make this change to fix an existing test that was broken by this change:
```
test/dialects/dialects_test.py::test__dialect__base_file_parse[bigquery-select_struct.sql]
```

File: test/fixtures/parser/ansi/double_dot.yml
Comment (barrywhart) [inline test/fixtures/parser/ansi/double_dot.yml:19]: Why'd this change? I don't know.

File: test/fixtures/parser/ansi/transactions.yml
Comment (barrywhart) [inline test/fixtures/parser/ansi/transactions.yml:1]: Why'd this change? I don't know.

File: test/fixtures/parser/mysql/force_index.yml
Comment (barrywhart) [inline test/fixtures/parser/mysql/force_index.yml:1]: Why'd this change? I don't know.

File: test/fixtures/parser/mysql/use_index.yml
Comment (barrywhart) [inline test/fixtures/parser/mysql/use_index.yml:1]: Why'd this change? I don't know.

**Comment by barrywhart** [comment]
I'm putting this PR in "Draft" status because of the likely impact on lint rules that look for `function_name` in the parse tree.

File: src/sqlfluff/core/rules/std/L010.py
Comment (barrywhart) [inline src/sqlfluff/core/rules/std/L010.py:176]: Introducing this new overridable function allows rule L030, which handles function names, to continue working even though a parsed `function_name` now includes an additional level.

File: src/sqlfluff/dialects/dialect_ansi.py
Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_ansi.py:752]: This new segment essentially replaces the old regex-based `FunctionNameSegment` so that rules looking for `function_name` in the parse tree are guaranteed to find it regardless of quoting.

File: src/sqlfluff/dialects/dialect_ansi.py
Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_ansi.py:254]: Renaming this now that `FunctionNameSegment` is a class.

File: src/sqlfluff/dialects/dialect_bigquery.py
Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_bigquery.py:178]: Renaming this to stay aligned with `dialect_ansi.py`.

File: src/sqlfluff/core/rules/std/L010.py
Comment (alanmcruickshank) [reply src/sqlfluff/core/rules/std/L010.py:174]: Can you add a note to that effect in the docstring for `_get_fix()`. Logic makes sense, but I assume future versions of ourselves will forget.

File: src/sqlfluff/core/rules/std/L030.py
Comment (alanmcruickshank) [inline src/sqlfluff/core/rules/std/L030.py:44]: Likewise I think this function needs some explanation in the docstring. It's not obvious why we need to to this from the code.

File: src/sqlfluff/dialects/dialect_ansi.py
Comment (alanmcruickshank) [reply src/sqlfluff/dialects/dialect_ansi.py:766]: I think we should probably have `allow_gaps=False` no? Unless there can be spaces between the elements. Otherwise I wonder whether this could behave strangely.

File: test/fixtures/parser/ansi/transactions.yml
Comment (alanmcruickshank) [reply test/fixtures/parser/ansi/transactions.yml:1]: Probably CRLF > LF

File: test/fixtures/parser/bigquery/select_udf_quote_nothing.yml
Comment (alanmcruickshank) [inline test/fixtures/parser/bigquery/select_udf_quote_nothing.yml:13]: Hmmmmm, I wonder whether the `identifiers` should be _part of_ `function_name`, i.e. that it should parse as:

```
function:
- function_name:
  - identifier: gcpproject
  - dot: .
  - idenfifier: functions
  - dot: .
  - function_name_identifier: timestamp_parsing
- start_bracket: (
...
```

Otherwise it feels wierd to nest `function_name_identifier` within `function_name` if there's never anything in there with it.

File: test/fixtures/parser/mysql/force_index.yml
Comment (alanmcruickshank) [reply test/fixtures/parser/mysql/force_index.yml:1]: Probably CRLF > LF

File: test/fixtures/parser/mysql/use_index.yml
Comment (alanmcruickshank) [reply test/fixtures/parser/mysql/use_index.yml:1]: Probably CRLF > LF

**Review by alanmcruickshank — CHANGES_REQUESTED** [review]
Other than some documentation, there's only one thing I don't understand here...

I think either we should nest the other identifiers within `function_name` (as per my more detailed comment) _or_ we don't need to have a nested `function_name_identifier` and `function_name`. I don't think it makes sense to nest it and then never put anything in there with it. Or have I missed some thinking on this?

File: src/sqlfluff/core/rules/std/L010.py
Comment (barrywhart) [reply src/sqlfluff/core/rules/std/L010.py:174]: Will do. I'm surprised the linter didn't flag this.

File: src/sqlfluff/dialects/dialect_ansi.py
Comment (barrywhart) [reply src/sqlfluff/dialects/dialect_ansi.py:766]: I wasn't familiar with that property, but yes, that makes sense -- I assume it disallows white space? I'll add it.

File: test/fixtures/parser/bigquery/select_udf_quote_nothing.yml
Comment (barrywhart) [reply test/fixtures/parser/bigquery/select_udf_quote_nothing.yml:13]: I'll look into this after I make the simpler changes. This PR had involved some "trial and error" and did kind of end up in a weird place. The main challenge, as I recall, was to allow hyphens in function names at the parse level, since it apparently couldn't be done at the lexer level.

Comment (barrywhart) [reply test/fixtures/parser/bigquery/select_udf_quote_nothing.yml:13]: Never mind about the hyphens -- I'm confusing this with an earlier PR for project names in the `FROM` clause, which *did* allow non-quoted hyphens. That one was a little odd. I think a more straightforward approach should work here.

**Comment by barrywhart** [comment]
@alanmcruickshank: I made the change you suggested re: `function_name` and `function_name_identifier`. I think I hadn't done this initially because I was worried about possible bad interactions with `CreateFunctionStatementSegment`, which also uses `FunctionNameSegment`, but after a second look, it should be fine to change how function names work there as well.

This change caused me to notice one remaining oddity about function names -- BigQuery allows quoting one or more dotted function name parts together, e.g.:
```
SELECT
    `another-gcp-project`.`functions.timestamp_parsing` (log_tbl.first_move) AS first_move
FROM
    `gcp-project.data.year_2021` AS log_tbl
```

In cases like this, the parse tree respects the quoting, so the quoted `.` becomes part of the identifier rather than a parse tree element. I think this is okay but may someday require helper functions in order for lint rules to properly interpret function names, since the same **logical** name can be represented in various ways.
```
[11](1, 2, 5)       |                    function:
[11](1, 2, 5)       |                        function_name:
[11](1, 2, 5)       |                            identifier:                       '`another-gcp-project`'
[32](1, 2, 26)      |                            dot:                              '.'
[33](1, 2, 27)      |                            identifier:                       '`functions.timestamp_parsing`'
```

**Comment by barrywhart** [comment]
@alanmcruickshank: I made your suggested change `FunctionNameSegment`, so this is ready for another review. Thanks!

File: src/sqlfluff/dialects/dialect_ansi.py
Comment (alanmcruickshank) [inline src/sqlfluff/dialects/dialect_ansi.py:757]: Small nit - I think this docstring is out of date now. `FunctionNameSegment` explicitly allows the option to **include** project, schema etc, and this says it excludes it.

**Review by alanmcruickshank — CHANGES_REQUESTED** [review]
One small nit on the docstring for FunctionNameSegment. Otherwise good to go! 👍

File: src/sqlfluff/dialects/dialect_ansi.py
Comment (barrywhart) [reply src/sqlfluff/dialects/dialect_ansi.py:757]: Ah yes -- I'll fix that now.

**Comment by barrywhart** [comment]
@alanmcruickshank: I fixed the docstring.

**Review by NiallRees — APPROVED** [review]

**Comment by barrywhart** [comment]
@alanmcruickshank: Can you take another look at this PR? Niall approved it, but you had requested a change (which I have now made). I need your approval to merge. 🙏🏽

**Review by NiallRees — APPROVED** [review]

**Comment by NiallRees** [comment]
I hadn't realised I'd already approved this! 🙈

**Comment by barrywhart** [comment]
@NiallRees: No worries! Yes, I can't merge it until Alan approves it, since earlier he requested changes.

**Review by alanmcruickshank — APPROVED** [review]
👍 Love it

**Pull request merged**
